File: 0152-maximum-product-subarray/0152-maximum-product-subarray.ts
function maxProduct(nums: number[]): number {
    let ret: number = nums[0];
    let min: number=1;
    let max: number=1;
    
    for(const n of nums){
    //nums.forEach((n) => {
        const preVal = [min*n, max*n, n];
        max = Math.max(...preVal);
        min = Math.min(...preVal);
        
        ret = Math.max(max, ret);
    //});
    }
    
    return ret;
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cs
public class Solution {
    public int MaxProduct(int[] nums) {
        int ret = nums[0];
        int max = 1, min = 1, prevMax = 1;
        
        foreach(int n in nums)
        {
            prevMax = max;
            max = Math.Max(max*n, Math.Max(min*n, n));
            min = Math.Min(prevMax*n, Math.Min(min*n, n));
            ret = Math.Max(max, ret);
            
            Console.WriteLine($" {max}:{min}:{ret}");
        }
        
        return ret;
    }
}